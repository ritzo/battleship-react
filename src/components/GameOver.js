import React from 'react';
import { Link } from 'react-router-dom';

import './GameOver.css';

class GameOver extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div>
        <h1>GameOver</h1>
        <Link to="/">Play again</Link>
      </div>
    );
  }
}

export default GameOver;
