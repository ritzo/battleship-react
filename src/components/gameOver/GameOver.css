.game-over {
    text-align: center;
}

.game-over .play {
    padding: 10px;
    width: 100%;
}
