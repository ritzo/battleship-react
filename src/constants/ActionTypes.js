// Actions related to the game setup
export const SETUP_BOARD = 'SETUP_BOARD';
export const SET_NAME = 'SET_NAME';
export const ADD_SHIP = 'ADD_SHIP';
export const PLAY = 'PLAY';
// Actions related to the game itself
export const SURRENDER = 'SURRENDER';
export const SHOOT = 'SHOOT';
export const OPONENT_TURN = 'OPONENT_TURN';
