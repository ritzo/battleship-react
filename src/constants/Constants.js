export const STATES = {
  OPEN: 'open',
  PLAYING: 'playing',
  FINISHED: 'finished',
};

export const SHIPS = [2, 3, 3, 3, 4];

export const POSITIONS = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
};

export const PLAYERS = {
  PLAYER: 'player',
  OPONENT: 'cpu',
};

export const SQUARE_STATES = {
  EMPTY: 'empty',
  INVALID: 'invalid',
  SHIP: 'ship',
  SHOOT_MISS: 'shoot_miss',
  SHOOT_SHIP: 'shoot_ship',
};

export const BOARD_TYPES = {
  EDITION: 'edition',
  PLAYER: 'player',
  OPONENT: 'oponent',
};
