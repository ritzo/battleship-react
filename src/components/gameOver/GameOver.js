import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';

import { STATES, PLAYERS } from '../../constants/Constants';

import './GameOver.css';

class GameOver extends React.Component {
  constructor(props) {
    super(props);

    this.actions = props.actions;
  }

  handlePlayAgainClick() {
    this.actions.setupBoard();
  }

  renderInformation() {
    const { winner, surrendered } = this.props;
    const isWinner = winner === PLAYERS.PLAYER;

    if (isWinner) {
      return (
        <p>You won, congratulations!</p>
      );
    }
    return (
      <div>
        <p>You lost, better luck next time</p>
        <p>{ (surrendered ? 'You ran away' : 'You fought bravely') }</p>
      </div>
    );
  }

  render() {
    const {
      state,
    } = this.props;

    if (state === STATES.OPEN) {
      return <Redirect to="/" />;
    }

    return (
      <div className="game-over">
        <h1>Game Over</h1>

        { this.renderInformation() }

        <button type="button" className="play" onClick={() => this.handlePlayAgainClick()}>Play again</button>
      </div>
    );
  }
}

GameOver.propTypes = {
  state: PropTypes.string.isRequired,
  winner: PropTypes.string.isRequired,
  surrendered: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    setupBoard: PropTypes.func.isRequired,
  }).isRequired,
};

export default GameOver;
