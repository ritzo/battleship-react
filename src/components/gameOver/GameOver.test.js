import React from 'react';
import { shallow } from 'enzyme';

import GameOver from './GameOver';

import { PLAYERS, STATES } from '../../constants/Constants';

describe('GameOver', () => {
  it('Snapshot', () => {
    const state = STATES.FINISHED;
    const winner = PLAYERS.PLAYER;
    const surrendered = false;
    const actions = {
      setupBoard() {},
    };

    const enzymeWrapper = shallow(
      <GameOver state={state} winner={winner} surrendered={surrendered} actions={actions} />,
    );

    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('Should redirect to home when state is "open"', () => {
    const state = STATES.OPEN;
    const winner = PLAYERS.PLAYER;
    const surrendered = false;
    const actions = {
      setupBoard() {},
    };

    const enzymeWrapper = shallow(
      <GameOver state={state} winner={winner} surrendered={surrendered} actions={actions} />,
    );

    expect(enzymeWrapper.find('Redirect')).toBeDefined();
    expect(enzymeWrapper.find('Redirect').props().to).toBe('/');
  });

  it('Should redirect to home when state is "playing"', () => {
    const state = STATES.PLAYING;
    const winner = PLAYERS.PLAYER;
    const surrendered = false;
    const actions = {
      setupBoard() {},
    };

    const enzymeWrapper = shallow(
      <GameOver state={state} winner={winner} surrendered={surrendered} actions={actions} />,
    );

    expect(enzymeWrapper.find('Redirect')).toBeDefined();
    expect(enzymeWrapper.find('Redirect').props().to).toBe('/game');
  });

  it('Should trigger action "setupBoard" on button click', () => {
    const state = STATES.FINISHED;
    const winner = PLAYERS.PLAYER;
    const surrendered = false;
    const actions = {
      setupBoard: jest.fn(),
    };

    const enzymeWrapper = shallow(
      <GameOver state={state} winner={winner} surrendered={surrendered} actions={actions} />,
    );

    enzymeWrapper.find('button').simulate('click');

    expect(actions.setupBoard).toHaveBeenCalled();
  });
});
