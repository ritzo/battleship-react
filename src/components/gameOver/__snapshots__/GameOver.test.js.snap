// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOver Snapshot 1`] = `
<div
  className="game-over"
>
  <h1>
    Game Over
  </h1>
  <p>
    You won, congratulations!
  </p>
  <button
    className="play"
    onClick={[Function]}
    type="button"
  >
    Play again
  </button>
</div>
`;
